### ISSUE TITLE
Linking the first post in a long thread appears to hide the thread

### ISSUE THREAD

**Issue by mcclure** [mastodon/mastodon#24676]
### Steps to reproduce the problem

Follow this link while logged in to Mastodon.social.

https://mastodon.social/@mcc/108199886340178151

<a href="https://user-images.githubusercontent.com/277318/234689544-98696c85-f778-497c-91af-06152743b1a8.png"><img width=400 src="https://user-images.githubusercontent.com/277318/234689544-98696c85-f778-497c-91af-06152743b1a8.png"></a>

This thread is quite long and has many replies (I did a diary kind of thing nearly every day for a year) (no, I'm no longer updating it, I made a new thread).

If you are logged in to Mastodon, and you follow a link *directly* to this post, you will not see any replies. If you wait for a very long time, longer than anyone would think to wait— it seems to be around a minute— the rest of the thread appears. There is no indication in the meantime that anything is loading.

If you are logged out, the thread loads as expected.

If you load a *late* post in the thread, for example the [end](https://mastodon.social/@mcc/110249807470595996), the thread loads as expected. If you are late in the thread and you highlight the first post, it comes up immediately without the missing context. 

If instead of mastodon.social you load from a [glitchsoc instance, or a 3.x instance](https://mastodon.social/@Doephin@awoo.space/110266895657917777), the thread loads as expected.

### Expected behaviour

My expectation is I should see as much of the thread when loading the first post in a window as when loading the last post. However the bigger/biggest problem here is that there is no indication a load is occurring. If there were a throbber, or a "click for replies" link, this would not be so bad. But the "failure" is silent.

### Actual behaviour

If I link someone this thread, they will think it is a single post.

### Specifications

Mastodon.social is running 4.1.2. I tested this in both advanced mode and "regular" mode. I tested it with my browser extensions turned off and saw the same behavior.

**Comment by ClearlyClaire** [comment]
Related to #20345

I think there are two issues here:
- the whole context is loaded at once (in this case, the REST API reply is nearly 3MB of JSON!) which takes a very long time, both server-side and client-side. Adding pagination to the context API and related UI is planned but it's pretty complex
- there is no loading indicator of any kind while the context is being loaded; this should be far easier to fix

> If you are logged in to Mastodon, and you follow a link _directly_ to this post, you will not see any replies. If you wait for a very long time, longer than anyone would think to wait— it seems to be around a minute— the rest of the thread appears. There is no indication in the meantime that anything is loading.

Yeah, loading the context is done in a single query, which can load thousands of replies and ancestors, we do need some kind of pagination—both in the API and UI—but it's a pretty complex task, especially since we want to represent a “flat” view as much as possible. The fact that there is no loading indicator makes the issue worse for sure.

> If you are logged out, the thread loads as expected.

This is explained by the logged out view having drastically reduced limits on the loaded context (#20355)
, so logged out visitors only get a fraction of the thread and replies, which is not ideal either because there is no pagination logic yet or any indication, but at least allows some of the context to load fast.

> If you load a _late_ post in the thread, for example the [end](https://mastodon.social/@mcc/110249807470595996), the thread loads as expected. If you are late in the thread and you highlight the first post, it comes up immediately without the missing context.

In general, the later you are in a thread, the least branching descendants there are, so less posts to load.

> If instead of mastodon.social you load from a [glitchsoc instance, or a 3.x instance](https://mastodon.social/@Doephin@awoo.space/110266895657917777), the thread loads as expected.

I don't think there is any meaningful difference in the code there, the most likely difference is that those instances are not aware of nearly as many replies.

**Comment by nemobis** [comment]
Not sure it's the same issue, but another example is https://mastodon.social/@Mastodon/110294411952997299 which doesn't show any replies when loaded while logged in. When visited logged out, a few dozens replies are shown.

**Comment by sk22** [comment]
> Adding pagination to the context API and related UI is planned but it's pretty complex

maybe there could be a depth limit for descendants (both a hard-coded default and a query parameter)? i think (though i might be wrong, i don't know how this is implemented at all) that this might be rather easy to implement, at least in comparison to pagination (which is also needed; like, who's gonna stop someone to spam countless direct replies to a post, adding no depth at all?)

i think the depth limit would fix the thread problem while not posing a big frontend problem - the last status being displayed would still have a reply count, so the frontend could know that it needs to fetch the content for that last post to keep loading (or display a "load more" button) - or alternatively, the user might see the reply count and just open that last post directly.

**Comment by simonw** [comment]
I just ran into this too. I wanted to link to this thread: https://mastodon.social/@mhoye/111335603309582734

But that URL gives me just the first post:

<img width="830" alt="CleanShot 2023-11-01 at 15 47 02@2x" src="https://github.com/mastodon/mastodon/assets/9599/a02d2492-5af4-46c2-ae17-ff8167ae9678">

Linking to the second post in the thread DOES display the thread: https://mastodon.social/@mhoye/111336017090790537

<img width="830" alt="CleanShot 2023-11-01 at 15 48 07@2x" src="https://github.com/mastodon/mastodon/assets/9599/7c3755a0-8538-42c3-8826-233c54deb25d">

**Comment by ClearlyClaire** [comment]
> I just ran into this too. I wanted to link to this thread: https://mastodon.social/@mhoye/111335603309582734

This is not exactly the same issue, but this is related: this is the result of the aggressive cut-off we do when logged-out to avoid the issue described in this thread.

**Comment by mcclure** [comment]
> > I just ran into this too. I wanted to link to this thread: https://mastodon.social/@mhoye/111335603309582734
> 
> This is not exactly the same issue, but this is related: this is the result of the aggressive cut-off we do when logged-out to avoid the issue described in this thread.

Oh, If I understand what's happening correctly this seems very bad… I've been very publicly harsh against MuskTwitter for doing this exact thing to logged out users.

Would it be possible when executing the cut-off to display some sort of notice, or "more" link, that some amount of information has been redacted? It seems showing a logged out user an incorrect representation of the thread with no notice something has been silently changed is bad (and possibly dangerous) behavior.

**Comment by ClearlyClaire** [comment]
> Would it be possible when executing the cut-off to display some sort of notice, or "more" link, that some amount of information has been redacted?

Unfortunately, doing that properly requires implementing pagination logic, which is not conceptually easy, and would require significant changes in the backend, API, and UI.

**Comment by mcclure** [comment]
I don't think I agree it would be that complicated. Consider two things:

- Even simply adding the words "log in to see rest of thread" would be a big improvement. I cannot believe that would require a full pagination logic; I don't know your exact architecture, but surely just a database column for "has thread replies" could be set when a thread reply is created.

    Threads often represent single thoughts cut across multiple posts. If you have bandwidth/resource limitations then those are the limitations you're working with, but not *disclosing* that content has been hidden could create misunderstandings and social problems. A post intended to be read as a thread of 2 posts could have its meaning changed if you remove the second post. Posts often have corrections to their original content posted as replies. Etc. If you at least disclose some content has been hidden, then a reader looking at a post will at least understand they might not have the entire story.

- My tests match those of simonw's— that this is a problem *exclusive* to the root post in a thread. If I view a non-initial post in a thread on mastodon.com, the entire thread displays. (Moreover, even with an initial post, it displays replies— just not replies by the original poster.) So even if you *do* include a link, you don't need full pagination to address the problem with current pagination— you only need to handle the special case of the first post in a thread which has at least one thread reply, and all you have to do is link to the second post in the thread. This will still save at least some bandwidth because not all readers will click to view the second post.

**Comment by simonw** [comment]
Yeah, I would hope that the check for "does a post exist in the database which is a reply to the-post-that-is-being-displayed and is written by the same author" would be inexpensive - that should be enough to identify if something is part of a thread or not.

In SQL terms, something like this:
```sql
select 1 from statuses where in_reply_to_id = :post_id and account_id = :account_id limit 1
```
